Lombiq.Tests.UI/Services/UITestExecutionSession.cs

```diff
@@ -221,16 +227,33 @@ private Exception PrepareAndLogException(Exception ex)
         return ex;
     }
 
-    private async Task CreateFailureDumpAsync(
-        Exception ex,
+    private async Task FailureTestDumpProcessAsync(string dumpContainerPath, Exception ex)
+    {
+        if (_context == null) return;
+
+        var debugInformationPath = GetDebugInformationPath(dumpContainerPath);
+
+        if (_context.IsBrowserRunning) await CaptureBrowserUsingDumpsAsync(debugInformationPath);
+
+        if (_dumpConfiguration.CaptureAppSnapshot) await CaptureAppSnapshotAsync(dumpContainerPath);
+
+        CaptureMarkupValidationResults(ex, debugInformationPath);
+    }
+
+    private async Task CreateTestDumpAsync(
         string dumpRootPath,
         int retryCount,
-        IDictionary<string, IFailureDumpItem> failureDumpContainer)
+        IDictionary<string, ITestDumpItem> testDumpContainer,
+        Func<string, Task> additionalDumpProcess = null)
     {
-        if (!_dumpConfiguration.CreateFailureDump) return;
+        if (!_dumpConfiguration.CreateTestDump ||
+            ((testDumpContainer == null || !testDumpContainer.Any()) && additionalDumpProcess == null))
```

Comment:
imo it's easier to understand with less parentheses
```suggestion
        if (!_dumpConfiguration.CreateTestDump ||
            (testDumpContainer?.Any() != true && additionalDumpProcess == null))
```